File: Samples/HelloBlinkyBackground/CS/StartupTask.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net.Http;
using Windows.ApplicationModel.Background;
using Windows.Devices.Gpio;
using Windows.System.Threading;
using System.Diagnostics;

namespace BlinkyHeadlessCS
{
    public sealed class StartupTask : IBackgroundTask
    {
        BackgroundTaskDeferral deferral;
        private GpioPinValue value = GpioPinValue.High;
        private const int LED_PIN = 5;
        private GpioPin pin;
        private ThreadPoolTimer timer;
        IBackgroundTaskInstance _taskInstance = null;

        BackgroundTaskCancellationReason _cancelReason = BackgroundTaskCancellationReason.Abort;
        volatile bool _cancelRequested = false;

        public void Run(IBackgroundTaskInstance taskInstance)
        {
            Debug.WriteLine("Background " + taskInstance.Task.Name + " Starting...");
            taskInstance.Canceled += new BackgroundTaskCanceledEventHandler(OnCanceled);

            deferral = taskInstance.GetDeferral();
            _taskInstance = taskInstance;

            InitGPIO();
            timer = ThreadPoolTimer.CreatePeriodicTimer(Timer_Tick, TimeSpan.FromMilliseconds(500));
            
        }
        private void InitGPIO()
        {
            pin = GpioController.GetDefault().OpenPin(LED_PIN);
            pin.Write(GpioPinValue.High);
            pin.SetDriveMode(GpioPinDriveMode.Output);
        }

        private void Timer_Tick(ThreadPoolTimer timer)
        {
            if (_cancelRequested == false)
            {
                value = (value == GpioPinValue.High) ? GpioPinValue.Low : GpioPinValue.High;
                pin.Write(value);
            } else
            {
                timer.Cancel();
                //
                // Indicate that the background task has completed.
                //
                deferral.Complete();
            }
            
        }

        private void OnCanceled(IBackgroundTaskInstance sender, BackgroundTaskCancellationReason reason)
        {
            //
            // Indicate that the background task is canceled.
            //
            _cancelRequested = true;
            _cancelReason = reason;

            Debug.WriteLine("Background " + sender.Task.Name + " Cancel Requested...");
        }

    }
}

File: Samples/HelloBlinkyBackground/VB/StartupTask.vb
' Copyright (c) Microsoft. All rights reserved.

Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Net.Http
Imports Windows.ApplicationModel.Background
Imports Windows.Devices.Gpio

' The Background Application template is documented at http://go.microsoft.com/fwlink/?LinkID=533884&clcid=0x409

Public NotInheritable Class StartupTask
	Implements IBackgroundTask

    Dim pin As GpioPin
    Dim deferral As BackgroundTaskDeferral
    Dim timer As Windows.System.Threading.ThreadPoolTimer
    'Dim lightOn As Boolean
    Dim value As GpioPinValue = GpioPinValue.High

    Dim _cancelReason = BackgroundTaskCancellationReason.Abort
    Dim _cancelRequested As Boolean = False

    Public Sub Run(taskInstance As IBackgroundTaskInstance) Implements IBackgroundTask.Run

        Debug.WriteLine("Background " + taskInstance.Task.Name + " Starting...")
        AddHandler taskInstance.Canceled, AddressOf OnCanceled

        deferral = taskInstance.GetDeferral()
        pin = GpioController.GetDefault().OpenPin(5)
        pin.SetDriveMode(GpioPinDriveMode.Output)
        ' lightOn = False
        Windows.System.Threading.ThreadPoolTimer.CreatePeriodicTimer(AddressOf Tick, TimeSpan.FromMilliseconds(500))

    End Sub

    Public Sub Tick(timer As Windows.System.Threading.ThreadPoolTimer)

        If (_cancelRequested = False) Then
            value = If(value = GpioPinValue.High, GpioPinValue.Low, GpioPinValue.High)
            pin.Write(value)
        Else
            timer.Cancel()
            ' Indicate that the background task has completed.
            deferral.Complete()
        End If

    End Sub

    Private Sub OnCanceled(sender As IBackgroundTaskInstance, reason As BackgroundTaskCancellationReason)

        _cancelRequested = True
        _cancelReason = reason

        Debug.WriteLine("Background " + sender.Task.Name + " Cancel Requested...")
    End Sub


End Class
